File: lua/mashless/recorder.lua
-- mashless.recorder
-- Captures normal/visual-mode keystrokes (with cursor positions) for the
-- current Neovim session. Insert-mode keys are ignored except arrow keys.

local M = {}

-- Arrow keys, used both to flag normal-mode arrow use and insert-mode arrows.
local ARROWS = {
  ['<Up>'] = true,
  ['<Down>'] = true,
  ['<Left>'] = true,
  ['<Right>'] = true,
}

-- Tokens whose surrounding line text we keep, so the analyzer can suggest
-- f/t/search targets for horizontal motions.
local KEEP_TEXT = {
  ['l'] = true,
  ['h'] = true,
  ['<Right>'] = true,
  ['<Left>'] = true,
  ['w'] = true,
  ['b'] = true,
  ['e'] = true,
  ['W'] = true,
  ['B'] = true,
  ['E'] = true,
}

local ns = vim.api.nvim_create_namespace('mashless_on_key')

local function now_ms()
  return (vim.uv or vim.loop).now()
end

-- Reduce Neovim's mode string to a coarse bucket.
local function classify(mode)
  local c = mode:sub(1, 1)
  if c == 'n' then
    return 'normal'
  elseif c == 'v' or c == 'V' or c == '\22' or c == 's' or c == 'S' or c == '\19' then
    return 'visual'
  elseif c == 'i' then
    return 'insert'
  end
  return 'other'
end

local function new_session()
  return {
    start_time = os.time(),
    start_ms = now_ms(),
    keylog = {}, -- ordered list of { tok, mode, line, col, t, text? }
    total_keys = 0, -- normal + visual keystrokes
    insert_keys = 0,
    insert_arrows = 0,
    files = {}, -- set: absolute path -> true
    last_pos = { 1, 0 }, -- most recent cursor position, kept fresh by CursorMoved
  }
end

M.session = nil

-- Begin recording. Safe to call once per session.
function M.start()
  M.session = new_session()
  local s = M.session

  local first = vim.fn.expand('%:p')
  if first ~= '' then
    s.files[first] = true
  end

  vim.on_key(function(key, typed)
    -- Recording must never break the editor: swallow every error.
    pcall(function()
      local k = typed
      if k == nil or k == '' then
        k = key
      end
      if k == nil or k == '' then
        return
      end

      local tok = vim.fn.keytrans(k)
      local kind = classify(vim.api.nvim_get_mode().mode)

      if kind == 'insert' then
        s.insert_keys = s.insert_keys + 1
        if ARROWS[tok] then
          s.insert_arrows = s.insert_arrows + 1
        end
        return
      end

      if kind ~= 'normal' and kind ~= 'visual' then
        return -- cmdline, terminal, etc.
      end

      s.total_keys = s.total_keys + 1

      -- on_key fires *before* the key is processed, so this is the cursor
      -- position the motion starts from.
      local pos = { 1, 0 }
      local ok, p = pcall(vim.api.nvim_win_get_cursor, 0)
      if ok then
        pos = p
      end

      local entry = {
        tok = tok,
        mode = kind,
        line = pos[1],
        col = pos[2],
        t = now_ms(),
      }
      if KEEP_TEXT[tok] then
        local okl, ln = pcall(vim.api.nvim_get_current_line)
        if okl then
          entry.text = ln
        end
      end
      table.insert(s.keylog, entry)
    end)
  end, ns)
end

-- Stop recording (detach the on_key callback).
function M.stop()
  vim.on_key(nil, ns)
end

return M

File: src/recorder.rs
//! Session state. The Lua shim forwards every keystroke (plus cursor and
//! buffer events) over msgpack-RPC; this module folds those events into the
//! session the analyzer later consumes. It is the Rust port of the old
//! `mashless.recorder` — with the crucial difference that key *capture* still
//! happens in Lua (`vim.on_key`), since that API has no RPC binding.

use std::collections::BTreeSet;

use chrono::{DateTime, Local};

/// Arrow keys, flagged both in normal mode and (separately) in insert mode.
pub const ARROWS: [&str; 4] = ["<Up>", "<Down>", "<Left>", "<Right>"];

pub fn is_arrow(tok: &str) -> bool {
    ARROWS.contains(&tok)
}

/// Tokens whose surrounding line text we keep, so the analyzer can suggest
/// `f`/`t`/search targets for horizontal motions.
pub fn keep_text(tok: &str) -> bool {
    matches!(
        tok,
        "l" | "h" | "<Right>" | "<Left>" | "w" | "b" | "e" | "W" | "B" | "E"
    )
}

/// Coarse mode bucket, derived from Neovim's mode string.
#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Kind {
    Normal,
    Visual,
    Insert,
    Other,
}

/// Reduce Neovim's mode string to a coarse bucket. Mirrors the Lua `classify`.
pub fn classify(mode: &str) -> Kind {
    match mode.chars().next() {
        Some('n') => Kind::Normal,
        // v, V, <C-v> (0x16), s, S, <C-s> (0x13)
        Some('v') | Some('V') | Some('\u{16}') | Some('s') | Some('S') | Some('\u{13}') => {
            Kind::Visual
        }
        Some('i') => Kind::Insert,
        _ => Kind::Other,
    }
}

/// One recorded normal/visual-mode keystroke, with the cursor position it
/// started from. `text` is the line under the cursor, kept only for the tokens
/// in [`keep_text`].
pub struct KeyEntry {
    pub tok: String,
    pub line: i64,
    pub col: i64,
    pub text: Option<String>,
}

/// Tunables, supplied by the Lua shim at setup time.
pub struct Config {
    pub output_dir: String,
    pub min_keys: i64,
    pub vmin: i64,
    pub hmin: i64,
    pub xmin: i64,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            output_dir: ".".into(),
            min_keys: 20,
            vmin: 3,
            hmin: 5,
            xmin: 3,
        }
    }
}

/// Everything recorded for the current Neovim session.
pub struct Session {
    pub start: DateTime<Local>,
    /// Ordered stream of normal/visual keystrokes.
    pub keylog: Vec<KeyEntry>,
    /// Normal + visual keystrokes.
    pub total_keys: u64,
    /// Arrow keys pressed while in insert mode.
    pub insert_arrows: u64,
    /// Absolute paths of files visited this session.
    pub files: BTreeSet<String>,
    /// Most recent cursor position, kept fresh by CursorMoved.
    pub last_pos: (i64, i64),
}

impl Session {
    pub fn new() -> Self {
        Self {
            start: Local::now(),
            keylog: Vec::new(),
            total_keys: 0,
            insert_arrows: 0,
            files: BTreeSet::new(),
            last_pos: (1, 0),
        }
    }

    /// Fold one forwarded keystroke into the session. `mode` is Neovim's raw
    /// mode string at the moment the key was pressed; `text` is the current
    /// line (empty when the shim decided it wasn't worth sending).
    pub fn record_key(&mut self, tok: String, mode: &str, line: i64, col: i64, text: Option<String>) {
        match classify(mode) {
            Kind::Insert => {
                if is_arrow(&tok) {
                    self.insert_arrows += 1;
                }
            }
            Kind::Normal | Kind::Visual => {
                self.total_keys += 1;
                let text = if keep_text(&tok) {
                    text.filter(|t| !t.is_empty())
                } else {
                    None
                };
                self.keylog.push(KeyEntry {
                    tok,
                    line,
                    col,
                    text,
                });
            }
            Kind::Other => {}
        }
    }
}
